juju/model.py

```diff
@@ -43,32 +45,88 @@ def get_relation(self, relation_name, relation_id=None):
                 # errors, ideally integrating the error catching with Juju's mechanisms.
                 raise TooManyRelatedApps(relation_name, num_related, 1)
 
+    def _is_local_unit(self, unit_name):
+        return unit_name == self._local_unit_name
+
     def get_unit(self, unit_name):
-        return self._cache.get(Unit, unit_name)
+        return self._cache.get(CacheKey(Unit, unit_name), unit_name, self._is_local_unit(unit_name), self._backend, self._cache)
```

Comment:
I'm a bit confused by these changes, as this seems unrelated to the specific purpose of this PR. Is this merging other PRs, and if so can we get it clean with only the specific modifications for status?

Reply:
So, I was the one who suggested adding something like `CacheKey` so that we could start distinguishing the parameters that are actually supposed to be part of the key vs ones that just need to get sent on to the constructor but aren't necessarily part of the identity (we already have `cache` as a special case of this).  However, maybe that was a bit of a "swallow the spider to catch the fly" solution, in that the root issue was that this was adding so many new params to the constructor.

`unit_name` is now duplicated, so that's just an artifact of this particular approach.

The "is local unit" flag is only needed to enforce read-only-ness at call time, and if nothing else, we could always just not bother checking that at all in the Python code and let it fail at the Juju level. That feels a bit dirty, but this is a pretty edge case that indicates an error by the charm dev, so maybe it's ok if it lets us keep the code much cleaner.

The backend seems pretty necessary, since we want `model.unit.status` (for the local unit / app at least) to be able to call `status-get` and `status-set`. But, like the cache, it's definitely not part of the identity of the cache entity, so needs to be excluded from the key. However, unlike the `cache` param, it would still need to be passed in from the outside, which makes it more tricky to special-case.

Reply:
The backend indeed feels like a common need, much like the cache itself, which is now being provided into constructors without being part of the cache instance itself. So can just create the model cache with it, and have the cache always handing it as an argument into all the model classes.

For the backend to be able to do the proper actions considering the local unit name (and not trust juju to error), we should move the local unit name origin into the backend itself. After all, it is the one that knows about the local hook world. This will also avoid having to pass the unit name around even into the Model itself as we're doing today.

With those two ideas in place, we should be good to go, and actually simplify our logic a bit with the local unit name behavior. Do I miss some piece of the puzzle?